sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -104,24 +126,40 @@ def __init__(self, max_bins=256, subsample=int(2e5), random_state=None):
         self.subsample = subsample
         self.random_state = random_state
 
-    def fit(self, X, y=None):
+    def fit(self, X, support_missing_values=False):
         """Fit data X by computing the binning thresholds.
 
+        The first bin is reserved for missing values, if any.
+
         Parameters
         ----------
         X : array-like, shape (n_samples, n_features)
             The data to bin.
-        y: None
-            Ignored.
+        support_missing_values : ndarray of bool or bool, shape (n_features,)
+            For each feature, indicates whether the first bin should be
+            reserved for missing values. Note that inferring this from X would
+            be incorrect here, in general. The X that is passed here is the
+            training data (after the train/val split).
+            support_missing_values must be computed on the whole data
+            (before the split) so that the first bin is allocated if there are
+            missing values in the training data OR in the validation data.
```

Comment:
I guess I'm not sure if I'm convinced why we need to check the validation data and see if there are any `nan`s. To me that sounds like leakage and not okay?

Reply:
Imagine the following situation.

X is the whole data
X_train and X_val are training and validation data after train/val split

Let's say we reserve the first bin for missing values **iff** there are missing values in X_train (that's wrong)

Let's say there's no missing values in X_train: the first bin isn't reserved for missing values. It's just the bin for the smallest values.

If there are missing values in X_val, then in predict_binned(), the missing values will be considered to belong to the first bin. This is wrong, because the first bin wasn't reserved for missing values. Missing values are treated as the smallest values, and that's not something we want.

So this part: *we reserve the first bin for missing values **iff** there are missing values in X_train* isn't the proper way of doing things.

Instead, we reserve the first bin for missing values **iff there are missing values in X** (not X_train).

Note that the actual bin thresholds are of course only determined base on X_train (this wasn't the case before, and we fixed this recently in #13933). 

There is no data leak. If there are no missing values in X_train, then there is no knowledge to learn from the missing values anyway.

Reply:
If that's not convincing, think about it this way:

We **always** allocate the first bin for missing values... Except when there are no missing values in X.